src/draco/io/stl_encoder.cc

```diff
@@ -0,0 +1,141 @@
+#include "draco/io/file_writer_interface.h"
+
+namespace draco {
+
+StlEncoder::StlEncoder()
+    : out_buffer_(nullptr), in_point_cloud_(nullptr), in_mesh_(nullptr) {}
+
+Status StlEncoder::EncodeToFile(const Mesh &mesh, const std::string &file_name) {
+  in_mesh_ = &mesh;
+  std::unique_ptr<FileWriterInterface> file =
+      FileWriterFactory::OpenWriter(file_name);
+  if (!file) {
+    return Status(Status::IO_ERROR, "File couldn't be opened");
+  }
+  // Encode the mesh into a buffer.
+  EncoderBuffer buffer;
+  DRACO_RETURN_IF_ERROR(EncodeToBuffer(mesh, &buffer));
+  // Write the buffer into the file.
+  file->Write(buffer.data(), buffer.size());
+  return OkStatus();
+}
+
+Status StlEncoder::EncodeToBuffer(const Mesh &mesh, EncoderBuffer *out_buffer) {
+  in_mesh_ = &mesh;
+  out_buffer_ = out_buffer;
+  Status s = EncodeInternal();
+  in_mesh_ = nullptr; // cleanup
+  in_point_cloud_ = nullptr;
+  out_buffer_ = nullptr;
+  return s;
+}
+
+Status StlEncoder::EncodeInternal() {
+  // Write STL header.
+  std::stringstream out;
+  out << std::left << std::setw(80) << "generated using Draco";  // header is 80 bytes fixed size.
+  const std::string header_str = out.str();
+  buffer()->Encode(header_str.data(), header_str.length());
+
+  uint32_t num_faces = in_mesh_->num_faces();
+  buffer()->Encode(&num_faces, 4);
+
+  std::vector<uint8_t> stl_face;
+
+  const int pos_att_id =
+      in_mesh_->GetNamedAttributeId(GeometryAttribute::POSITION);
+  int normal_att_id =
+      in_mesh_->GetNamedAttributeId(GeometryAttribute::NORMAL);
+
+  if (pos_att_id < 0) {
+    return ErrorStatus("Mesh is missing the position attribute.");
+  }
+
+  // Ensure normals are 3 component. Don't encode them otherwise.
+  if (normal_att_id >= 0 &&
+      in_mesh_->attribute(normal_att_id)->num_components() != 3) {
+    normal_att_id = -1;
+  }
+
+  if (in_mesh_->attribute(pos_att_id)->data_type() !=
+      DT_FLOAT32) {
+    return ErrorStatus("Mesh position attribute is not of type float32.");
+  }
+
+  if (normal_att_id >= 0 &&
+      in_mesh_->attribute(normal_att_id)->data_type() != DT_FLOAT32) {
+    return ErrorStatus("Mesh normal attribute is not of type float32.");
+  }
+
+  uint16_t unused = 0;
+
+  if (in_mesh_) {
+    for (FaceIndex i(0); i < in_mesh_->num_faces(); ++i) {
+
+      const auto &f = in_mesh_->face(i);
+      const auto *const pos_att = in_mesh_->attribute(pos_att_id);
+
+      if (normal_att_id >= 0) {
+        const auto *const normal_att =
+            in_mesh_->attribute(normal_att_id);
+        buffer()->Encode(normal_att->GetAddress(normal_att->mapped_index(f[0])),
+                         normal_att->byte_stride());
+      } else {
+        // In case the source data does not contain normal vectors, we calculate them
+        // using the points of the trianlge face: cross(p2-p1, p3-p1)
+        // Finally we normalize the vector.
+        // TODO: can Eigen::Vector3f math be used here ?
+
+        std::array<float, 3> p1 =
```

Comment:
the below code can be probably simplified a bit:
```
Vector3f pos[3];
pos_att->GetMappedValue(f[0], &pos[0][0]);
pos_att->GetMappedValue(f[1], &pos[1][0]);
pos_att->GetMappedValue(f[2], &pos[2][0]);
const Vector3f norm = CrossProduct(pos[1] - pos[0], pos[2] - pos[0]);
norm.Normalize();
buffer()->Encode(norm.Data(), sizeof(float) * 3);
```

There may be some typos, I haven't tried the code.

Reply:
Awesome, much better, I saw Eigen::Vector3f but I wasn't sure if I was allowed to use it, it seemed to be behind some transcoder ifdefs. For some reason I missed Vector3f. Thank you!